DI system based on dataclasses

pauleveritt opened the pull request:
This adds an optional sub "package" at `wired.dataclasses`. If you have Python 3.7 installed or the 3.6 backport of dataclasses, then `wired.dataclasses` gives you a convenient way to register singletons and factories.

If you have `venusian` installed, you can use decorators instead of imperatively registering dataclasses.

The injection system has some interesting features, such as picking an attribute or a key off the injected object.

mmerickel commented on docs/conftest.py:
```diff
@@ -1,18 +1,4 @@
-        ],
-        pattern='index.rst',
-    ).pytest()
+    pass
```
it looks like this file should be removed

mmerickel commented on docs/dc/index.rst:
```diff
@@ -0,0 +1,89 @@
+Wouldn't it be nice if all you had to do was ask for what you wanted, and they system was responsible for providing it?
+This makes for far-less-coupled systems.
+Especially systems with more than two parties.
+Not just the web framework (or Sphix) and the site application, but plugins.
```
sphinx

mmerickel commented on docs/dc/simple_construction/index.rst:
```diff
@@ -0,0 +1,51 @@
+
+.. note::
+
+    We'll show later how a ``wired_factory`` staticmethod can let you write
```
static method or `staticmethod`

mmerickel commented on docs/dc/simple_construction/index.rst:
```diff
@@ -0,0 +1,51 @@
+
+But with ``wired.dataclasses``, no factory function is needed, as `wired.dataclasses` makes a factory for you:
+
+.. literalinclude:: configure.py
```
It looks like `configure.py` is included twice in this file and I don't think the contents is right as 1) it's not using/importing the `App` object and 2) it's making an unused container which probably shouldn't be there. Can you clarify for me what's going on here?

pauleveritt replied:
You're right, I removed the duplicate. The reason it doesn't seem to be used is a trick I did for doc purposes: the block at the bottom that is included in the docs, actually comes from the test.

But I didn't want the docs to show mega-long imports and I couldn't get relative working right.

I made a small adjustment for this, to show `configure.register`.

mmerickel commented on docs/dc/custom_factories/request.py:
```diff
@@ -0,0 +1,17 @@
+
+    # Manually make a Request and put in the container
+    request = Request(url=name)
+    container.set(request, Request)
```
this should use `register_singleton` after the changes in wired 0.2

pauleveritt replied:
Hmm, but i's value changes from request to request. I thought the container would be a good place to put such state.

mmerickel replied:
`container.register_singleton`, not `registry.register_singleton`. See the changelog in wired 0.2 about it. The `set` api only applies to a specific context in the container.

mmerickel commented on docs/dc/injected/request.py:
```diff
@@ -0,0 +1,17 @@
+
+    # Manually make a Request and put in the container
+    request = Request(url=name)
+    container.set(request, Request)
```
this should use register_singleton after the changes in wired 0.2

mmerickel commented on tests/dataclasses/integration/containers/request.py:
```diff
@@ -0,0 +1,17 @@
+
+    # Manually make a Request and put in the container
+    request = Request(url=name)
+    container.set(request, Request)
```
this should use register_singleton after the changes in wired 0.2

mmerickel commented on docs/dc/attributes/request.py:
```diff
@@ -0,0 +1,17 @@
+
+    # Manually make a Request and put in the container
+    request = Request(url=name)
+    container.set(request, Request)
```
this should use register_singleton after the changes in wired 0.2

mmerickel commented on docs/dc/injected/models.py:
```diff
@@ -0,0 +1,48 @@
+@dataclass
+class Greeter:
+    """ A basic greeter """
+    customer: Customer = injected(Context)
```
So the Context is acting as some form of proxy registered only to have a dynamic factory. An instance of Context is never actually returned by the `Context.wired_factory`. What can be done to make this more clear? Could we call `Context` a `DynamicCustomer` or something? Is a Context type actually necessary versus some sort of `injected(get_customer)`?

pauleveritt replied:
This is very much a place I could use architectural help. I've changed my mind frequently and still think the abstraction is wrong.

Pyramid has Context and it has Resource. But not every Context is in the resource tree (e.g. NotFound). I was trying to avoid conflating Context and Resource.

From a static typing perspective, I'm not sure how to represent it. Some of the `injected()` machinery it to allow a view to know "this is going to return a Customer".

On your last question, `context` is such a first-class part of Pyramid that I thought it critical to always be able to get an instance.

mmerickel replied:
The issue I was pointing out in this example is that `greeter.customer` is never an instance of `Context`. Rather, the `Context` here is used only for its `wired_factory` which never returns an instance of `Context`. It's really a proxy that returns a type of customer based on the values in another service. It might be better to call it a `DynamicCustomer` and do something like `customer: Customer = injected(DynamicCustomer)`. Or you could take a callable, versus a type:

```
def get_dynamic_customer(container):
    request = container.get(Request)
    return CUSTOMERS[request.url]

customer: Customer = injected(factory=get_dynamic_customer)
```

pauleveritt replied:
This is something I need to get right. I propose to move this conversation to a ticket where we can have a longer conversation. That ok?

mmerickel replied:
Yeah I think this probably needs to be discussed, the correct path forward is not super clear to me either.

mmerickel commented on src/wired/dataclasses/injector.py:
```diff
@@ -0,0 +1,110 @@
+                context: Context = container.get(Context)
+            except LookupError:
+                # It's ok to have context=None
+                pass
```
Can you help me understand the difference between this and `container.context`? It's not quite clicking for me yet.

pauleveritt replied:
Sometimes, in the process of injector-resolution, you have something asking for a different context than the container. (I doubt that's a satisfactory answer.)

pauleveritt replied:
Also, `register_dataclass` lets you [register a dataclass for a context](https://github.com/pauleveritt/wired/blob/dataclasses/tests/dataclasses/unit/test_registration.py#L62).

This is then used in the [decorator closure for the generated factory](https://github.com/pauleveritt/wired/blob/dataclasses/src/wired/dataclasses/registration.py#L25).

mmerickel replied:
Wired doesn't know anything about pyramid or requests. Each container does have a `context` attribute which is basically like `context = injected(ServiceContainer, attr='context')`. This would also be the instance of the context, versus the type that you pass to `register_dataclass`, but they are I'm pretty sure what you're looking at here.

pauleveritt replied:
I know wired is distinct from pyramid but I was trying to follow a Pyramid pattern which I thought wired did as well, meaning, there's something called a "context". 

How do I associate a particular context instance with a container? Is it `.bind`?

mmerickel replied:
`registry.create_container(context=...)` is the first way. Other than that you can later create other pseudo-containers that are bound to a context via `.bind` which returns a new container, sharing the same caches, but with a different bound context. `container.context` can change per-factory based on how the factory is registered. If this isn't clear I'll have to think of a better way to describe it - it's probably one of the most subtle parts of wired.

mmerickel replied:
I think the `context` arg should be removed here, and the `models.Context` idea dropped for this iteration.

pauleveritt replied:
I removed most of Context but left it in as something that people could ask for during injection.

I could change it so that Context isn't something that people can use in their dataclasses (factories.) It just seems...weird. I'm registering this factory for a context=FrenchCustomer but I can't ask for that context, I have to wire up some other mechanism to hand me the context, when it's right there in `container.context`.

mmerickel replied:
Who is registering the Context service? Why isn’t the injector just using container.context? You already made the ServiceContainer injectable so it’s just the context attr on that object. Doesn’t need to be separate really. The wired context is not a service it’s more like a scope.

mmerickel commented on src/wired/dataclasses/decorators.py:
```diff
@@ -0,0 +1,59 @@
+                context=self.context
+            )
+
+        attach(wrapped, callback)
```
`category='wired'`

pauleveritt replied:
Hmm, interesting, I had a feeling `category` was something I should know about. Can you explain the impact?

mmerickel replied:
It basically allows various libraries to only scan decorators they care about. It matters who is scanning the module (pyramid's configurator, versus this custom scanner, versus some other scanner) as each decorator expects certain attributes on the scanner. Thus you want to give the user control over which decorators apply to this scan. See also the default for the scanner is to pick up all categories which can exacerbate the problem.

https://github.com/Pylons/venusian/blob/72f46814884372cf70b343ef8bb5bdcb9c5919c5/venusian/__init__.py#L29-L32

mmerickel commented on src/wired/dataclasses/decorators.py:
```diff
@@ -0,0 +1,59 @@
+            for_ = self.for_ if self.for_ else cls
+            registry.register_singleton(instance, for_)
+
+        attach(wrapped, callback)
```
`category='wired'`

pauleveritt commented:
@mmerickel I believe I've addressed most of these points. There's still other work that can be done, but this might be a good checkpoint to decide if it does indeed go in the main package.

mmerickel commented on docs/dc/simple_construction/configure.py:
```diff
@@ -0,0 +1,9 @@
+
+
+def register(registry: ServiceRegistry):
+    container = registry.create_container()
```
this `container` is unused

mmerickel commented:
> @mmerickel I believe I've addressed most of these points. There's still other work that can be done, but this might be a good checkpoint to decide if it does indeed go in the main package.

@pauleveritt I'm definitely on board with merging this. I think the only remaining sticking points for me are the two unresolved comments (plus the one tiny comment I just added).

mmerickel commented:
@pauleveritt can you merge master into here and fix up the tests? There's some line warnings, 2 actual lines missing coverage and then a few in the tests.

pauleveritt commented:
I see a complaint about venusian's use of `imp` but I'm not sure the right way to silence that. I fixed all the flake8 things and the coverage thing. The stuff about black...I'll have to do some work to get my editor's choices to match black's.

pauleveritt commented:
@mmerickel Why does the coverage report look for coverage on stuff under /tests? The py37 Travis tox is failing because of that, for my Dummy fakes. What's the correct pattern, put those in a fixture so coverage will ignore them?

bertjwregeer commented:
There's a bug fix in venusian I need to review/pull and push a new version to fix that imp stuff. Ignore it for now IMHO.

Also, for some reason pytest under python 3.7 seems to find a lot more tests than pytest under other versions of python.

bertjwregeer commented:
```platform darwin -- Python 3.5.6, pytest-5.0.1, py-1.8.0, pluggy-0.12.0
cachedir: .tox/py35/.pytest_cache
rootdir: /Users/xistence/Projects/wired, inifile: setup.cfg, testpaths: src/wired, tests, docs
plugins: cov-2.7.1
collected 43 items```

vs

```platform darwin -- Python 3.7.0, pytest-5.0.1, py-1.8.0, pluggy-0.12.0
cachedir: .tox/py37/.pytest_cache
rootdir: /Users/xistence/Projects/wired, inifile: setup.cfg, testpaths: src/wired, tests, docs
plugins: cov-2.7.1
collected 106 items```

bertjwregeer commented:
Oh, that's because of the `conftest.py` files...

/me shakes fist at @pauleveritt

bertjwregeer commented:
@pauleveritt in the Pylons Project at least, we make sure that all of our `Dummy` fixtures all have coverage as well, so if there is code that is not covered in your dummy objects, remove the code.

pauleveritt commented:
I pushed tests for the dummies, back to 100% coverage.

mmerickel commented on tox.ini:
```diff
@@ -16,6 +16,10 @@ setenv =
 
+[testenv:py37]
+requires =
+    venusian
```
the better way to do this is to use tags on the `[testenv]`:

```ini
[testenv]
command = ...
requires =
    pip>=19.2.1
    py37: venusian
```

mmerickel replied:
Bert just replied with an even better solution which is the environment markers in the testing extra of setup.py.

pauleveritt replied:
The suggestion by @bertjwregeer is good, but we want to make it an optional dependency on Python 3.7. I believe his solution will install it automatically.

mmerickel replied:
Only for the testing extra, which is opt-in for the tests which is only used by `[testenv] extras = testing`. Of course don't add it to the default requirements.

bertjwregeer commented on tox.ini:
```diff
@@ -16,6 +16,10 @@ setenv =
 
+[testenv:py37]
+requires =
+    venusian
```
I would recommend adding this to `setup.py` with an environment marker for `python_version>=3.7`

pauleveritt replied:
Presumably you mean on this line:

```python
tests_require = ['pytest', 'pytest-cov', 'venusian', 'sybil']
```

How do I add venusian as a dependency with the python_version environment marker?

mmerickel replied:
`venusian; python_version >= 3.7`

mmerickel replied:
https://www.python.org/dev/peps/pep-0496/

mmerickel commented:
The docs are going to need a venusian dependency as well. That'll be a new extra and RTD as well as `[testenv:docs]` in tox.ini needs to be configured to install it.

pauleveritt commented:
Travis now passes. I made a new issue in #28 to remind us to fix Sphinx autodoc and the `__mro__` issue.

mmerickel commented on tox.ini:
```diff
@@ -66,14 +62,14 @@ deps =
+# build sdist/wheel
     python -m pep517.build .
-    # verify all is well
+# verify all is well
```
not sure why these are untabbed now, they were better at the same indentation level

pauleveritt replied:
I told PyCharm to reformat, which was dumb. Fixed.

mmerickel replied:
Not actually fixed, but I will do it myself after merging.

mmerickel commented:
thanks again @pauleveritt !
